enterprise/server/remote_execution/uffd/uffd.go

```diff
@@ -244,78 +240,48 @@ func (h *Handler) handle(ctx context.Context, memoryStore *copy_on_write.COWStor
 
-		guestPageAddr := pageStartAddress(guestFaultingAddr, pageSize)
-		if guestPageAddr < mapping.BaseHostVirtAddr {
-			// Make sure we only try to map addresses that fall within the valid
-			// guest memory ranges
-			guestPageAddr = mapping.BaseHostVirtAddr
+			uffdEvents = append(uffdEvents, event)
 		}
 
-		// Find the memory data in the store that should be used to handle the page fault
-		faultStoreOffset := guestMemoryAddrToStoreOffset(guestPageAddr, *mapping)
-
-		// To reduce the number of UFFD round trips, try to copy the entire
-		// chunk containing the faulting address
-		hostAddr, copySize, err := memoryStore.GetChunkStartAddressAndSize(uintptr(faultStoreOffset), false /*=write*/)
-		if err != nil {
-			return status.WrapError(err, "get backing page address")
-		}
-
-		relOffset := memoryStore.GetRelativeOffsetFromChunkStart(faultStoreOffset)
-		chunkStartOffset := faultStoreOffset - relOffset
-		destAddr := guestPageAddr - relOffset
-
-		// If copying the entire chunk would map data falling outside the valid
-		// guest memory range, only copy the valid parts of the chunk
-		//
-		// Check for data below the valid memory range
-		var invalidBytesAtChunkStart uintptr
-		if destAddr < mapping.BaseHostVirtAddr {
-			invalidBytesAtChunkStart = mapping.BaseHostVirtAddr - destAddr
-			destAddr = mapping.BaseHostVirtAddr
-			hostAddr += invalidBytesAtChunkStart
-			copySize -= int64(invalidBytesAtChunkStart)
-		}
-		// Check for data above the valid memory range
-		mappingEndAddr := mapping.BaseHostVirtAddr + mapping.Size
-		copyEndAddr := destAddr + uintptr(copySize)
-		var invalidBytesAtChunkEnd int64
-		if copyEndAddr > mappingEndAddr {
-			invalidBytesAtChunkEnd = int64(copyEndAddr - mappingEndAddr)
-			copySize -= invalidBytesAtChunkEnd
-		}
-
-		// Should never map a partial page at the end of the file, but just
-		// warn in this case for now.
-		if remainder := storeLength - int64(faultStoreOffset); remainder < int64(pageSize) {
-			log.CtxWarningf(ctx, "uffdio_copy range extends past store length")
-		}
-
-		_, err = h.resolvePageFault(uffd, uint64(destAddr), uint64(hostAddr), uint64(copySize))
-		if err != nil {
-			return err
+		for _, e := range uffdEvents {
+			if e.pagefault != nil {
+				// The memory location the VM tried to access that triggered the page fault.
+				guestFaultingAddr := e.pagefault.Address
+
+				mapping, err := guestMemoryAddrToMapping(uintptr(guestFaultingAddr), mappings)
+				if err != nil {
+					return err
+				}
+
+				// We can only handle page faults in units of pages, so align the faulting
+				// address to a page boundary. In other words, get the address of the
+				// start of the page containing the faulting address.
+				guestPageAddr := pageStartAddress(guestFaultingAddr, pageSize)
+				if guestPageAddr < mapping.BaseHostVirtAddr {
+					// Make sure we only try to map addresses that fall within the valid
+					// guest memory ranges
+					guestPageAddr = mapping.BaseHostVirtAddr
+				}
```

Comment:
wdyt about moving this logic into `handlePageFault` to reduce the amount of nested code in this func? and so the overall control flow is more clear (if e.pagefault { handle page fault }; if e.remove { handle remove }).

Reply:
To be honest I'm not sure why I left these lines out. Changed